purge-docker-cluster.yml

```diff
@@ -1,43 +1,447 @@
+    with_items: "{{ ceph_osd_docker_devices }}"
+    ignore_errors: true
+
+  - name: remove ceph osd container
+    docker:
+      image: "{{ ceph_osd_docker_username }}/{{ ceph_osd_docker_imagename }}:{{ ceph_osd_docker_image_tag }}"
+      name: "{{ ansible_hostname }}-osd-{{ item | regex_replace('/', '') }}"
+      state: absent
+    with_items: "{{ ceph_osd_docker_devices }}"
+    ignore_errors: true
+
+  - name: zap ceph osd disk
+    docker:
+      image: "{{ ceph_osd_docker_username }}/{{ ceph_osd_docker_imagename }}:{{ ceph_osd_docker_image_tag }}"
+      name: "{{ ansible_hostname }}-osd-zap-{{ item | regex_replace('/', '') }}"
+      net: host
+      pid: host
+      state: started
+      privileged: yes
+      env: "CEPH_DAEMON=zap_device,OSD_DEVICE={{ item }}"
+      volumes: "/var/lib/ceph:/var/lib/ceph,/etc/ceph:/etc/ceph,/dev:/dev,/run:/run"
+    with_items: "{{ ceph_osd_docker_devices }}"
+
+  - name: remove ceph osd zap disk container
+    docker:
+      image: "{{ ceph_osd_docker_username }}/{{ ceph_osd_docker_imagename }}:{{ ceph_osd_docker_image_tag }}"
+      name: "{{ ansible_hostname }}-osd-zap-{{ item | regex_replace('/', '') }}"
+      state: absent
+    with_items: "{{ ceph_osd_docker_devices }}"
+
+  - name: remove ceph osd service
+    file:
+        path: /etc/systemd/system/ceph-osd@.service
+        state: absent
+
+  - name: remove ceph osd image
+    docker_image:
+      state: absent
+      name: "{{ ceph_osd_docker_username }}/{{ ceph_osd_docker_imagename }}"
+      tag: "{{ ceph_osd_docker_image_tag }}"
+    tags:
+      remove_img
+
+
+- name: purge ceph mon cluster
+
+  hosts:
+    - mons
+
+  become: true
+
+  tasks:
+  - name: disable ceph mon service
+    service:
+      name: "ceph-mon@{{ ansible_hostname }}"
+      state: stopped
+      enabled: no
+    ignore_errors: true
+
+  - name: remove ceph mon container
+    docker:
+      image: "{{ ceph_mon_docker_username }}/{{ ceph_mon_docker_imagename }}:{{ ceph_mon_docker_image_tag }}"
+      name: "{{ ansible_hostname }}"
+      state: absent
+    ignore_errors: true
+
+  - name: remove restapi container
+    docker:
+      image: "{{ ceph_restapi_docker_username }}/{{ ceph_restapi_docker_imagename }}:{{ ceph_restapi_docker_image_tag }}"
+      name: "{{ ansible_hostname }}-ceph-restapi"
+      state: absent
+
+  - name: remove ceph mon service
+    file:
+        path: /etc/systemd/system/ceph-mon@.service
+        state: absent
+
+  - name: remove ceph mon image
+    docker_image:
+      state: absent
+      name: "{{ ceph_mon_docker_username }}/{{ ceph_mon_docker_imagename }}"
+      tag: "{{ ceph_mon_docker_image_tag }}"
+    tags:
+      remove_img
+
+
+- name: remove installed packages
+
+  hosts:
+    - all
+
+  become: true
+
+  tags:
+    with_pkg
+
+  tasks:
+  - name: stop docker service
+    service:
+      name: docker
+      state: stopped
+      enabled: no
+    when: not is_atomic
+
+  - name: remove pip and docker on ubuntu
```

Comment:
You're right - if we never installed it to begin with we definitely don't want to remove it. The problem is that we may have installed it to begin with so we'd be leaving extra stuff behind by not fully undoing all the steps we did. Unfortunately, we don't have an easy way to check if we installed it initially without leaving some sort of trail behind.

The idea is that we'd give the user the option of skipping these steps by passing the `--skip-tags=with_pkg` the way we do when we run the `ansible-playbook --skip-tags=with_pkg site-docker.yml` command. Unfavorably, it does require the user to know that these steps are going to happen and to use the tag to skip them.

On a side note: it'd be nice to have the opposite behavior supported in Ansible such that we _always_ skip these steps unless the user purposefully requests them. I'm thinking something like a `--no-skip-tags=with_pkg` option. I suppose the same thing could be done with a conditional that the user defines using `--extra-args`, but then you have to add the conditional check in each of these tasks unless you're using roles. I should look into seeing if anybody has made a feature request like this. Ok I looked and there is https://github.com/ansible/ansible/issues/11045 which would come in handy here.

Anyhow, what would you like to see happen here?